File: ShadowEditor.Server/OperatingAuthority.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using ShadowEditor.Server.CustomAttribute;

namespace ShadowEditor.Server
{
    /// <summary>
    /// 操作权限
    /// </summary>
    public static class OperatingAuthority
    {
        /// <summary>
        /// 管理员权限
        /// </summary>
        /// <remarks>只有管理员能操作，不应该被其他角色使用。</remarks>
        [Hidden]
        public const string ADMINISTRATOR = "Administrator";

        /// <summary>
        /// 只有登录用户才能具有的权限
        /// </summary>
        [Hidden]
        public const string LOGIN = "Login";

        /// <summary>
        /// 获取动画列表（不具备该权限无法查看动画列表）
        /// </summary>
        public const string LIST_ANIMATION = "List Animation";

        /// <summary>
        /// 添加动画
        /// </summary>
        public const string ADD_ANIMATION = "Add Animation";

        /// <summary>
        /// 编辑动画
        /// </summary>
        public const string EDIT_ANIMATION = "Edit Animation";

        /// <summary>
        /// 删除动画
        /// </summary>
        public const string DELETE_ANIMATION = "Delete Animation";

        /// <summary>
        /// 获取音频列表（不具备该权限无法查看音频列表）
        /// </summary>
        public const string LIST_AUDIO = "List Audio";

        /// <summary>
        /// 添加音频
        /// </summary>
        public const string ADD_AUDIO = "Add Audio";

        /// <summary>
        /// 编辑音频
        /// </summary>
        public const string EDIT_AUDIO = "Edit Audio";

        /// <summary>
        /// 删除音频
        /// </summary>
        public const string DELETE_AUDIO = "Delete Audio";

        /// <summary>
        /// 获取分类列表（不具有该权限无法进入分类窗口）
        /// </summary>
        public const string LIST_CATEGORY = "List Category";

        /// <summary>
        /// 保存分类
        /// </summary>
        public const string SAVE_CATEGORY = "Save Category";

        /// <summary>
        /// 删除分类
        /// </summary>
        public const string DELETE_CATEGORY = "Delete Category";

        /// <summary>
        /// 获取人物列表（不具备该权限无法进入人物列表）
        /// </summary>
        public const string LIST_CHARACTER = "List Character";

        /// <summary>
        /// 编辑人物
        /// </summary>
        public const string EDIT_CHARACTER = "Edit Character";

        /// <summary>
        /// 保存人物
        /// </summary>
        public const string SAVE_CHARACTER = "Save Character";

        /// <summary>
        /// 删除人物
        /// </summary>
        public const string DELETE_CHARACTER = "Delete Character";

        /// <summary>
        /// 获取贴图列表（不具有该权限无法查看贴图列表）
        /// </summary>
        public const string LIST_MAP = "List Map";

        /// <summary>
        /// 添加贴图
        /// </summary>
        public const string ADD_MAP = "Add Map";

        /// <summary>
        /// 编辑贴图
        /// </summary>
        public const string EDIT_MAP = "Edit Map";

        /// <summary>
        /// 删除贴图
        /// </summary>
        public const string DELETE_MAP = "Delete Map";

        /// <summary>
        /// 获取材质列表（不具有该权限无法查看材质列表）
        /// </summary>
        public const string LIST_MATERIAL = "List Material";

        /// <summary>
        /// 编辑材质
        /// </summary>
        public const string EDIT_MATERIAL = "Edit Material";

        /// <summary>
        /// 保存材质
        /// </summary>
        public const string SAVE_MATERIAL = "Save Material";

        /// <summary>
        /// 删除材质
        /// </summary>
        public const string DELETE_MATERIAL = "Delete Material";

        /// <summary>
        /// 获取模型列表（不具有该权限无法查看模型列表）
        /// </summary>
        public const string LIST_MESH = "List Mesh";

        /// <summary>
        /// 添加模型
        /// </summary>
        public const string ADD_MESH = "Add Mesh";

        /// <summary>
        /// 编辑模型
        /// </summary>
        public const string EDIT_MESH = "Edit Mesh";

        /// <summary>
        /// 删除模型
        /// </summary>
        public const string DELETE_MESH = "Delete Mesh";

        /// <summary>
        /// 获取粒子列表（不具有该权限无法查看粒子列表）
        /// </summary>
        public const string LIST_PARTICLE = "List Particle";

        /// <summary>
        /// 编辑粒子
        /// </summary>
        public const string EDIT_PARTICLE = "Edit Particle";

        /// <summary>
        /// 保存粒子
        /// </summary>
        public const string SAVE_PARTICLE = "Save Particle";

        /// <summary>
        /// 删除粒子
        /// </summary>
        public const string DELETE_PARTICLE = "Delete Particle";

        /// <summary>
        /// 获取预设体列表（不具有该权限无法查看预设体列表）
        /// </summary>
        public const string LIST_PREFAB = "List Prefab";

        /// <summary>
        /// 编辑预设体
        /// </summary>
        public const string EDIT_PREFAB = "Edit Prefab";

        /// <summary>
        /// 保存预设体
        /// </summary>
        public const string SAVE_PREFAB = "Save prefab";

        /// <summary>
        /// 删除预设体
        /// </summary>
        public const string DELETE_PREFAB = "Delete Prefab";

        /// <summary>
        /// 编辑场景
        /// </summary>
        public const string EDIT_SCENE = "Edit Scene";

        /// <summary>
        /// 保存场景
        /// </summary>
        public const string SAVE_SCENE = "Save Scene";

        /// <summary>
        /// 发布场景
        /// </summary>
        public const string PUBLISH_SCENE = "Publish Scene";

        /// <summary>
        /// 删除场景
        /// </summary>
        public const string DELETE_SCENE = "Delete Scene";

        /// <summary>
        /// 获取截图列表（不具有该权限无法查看截图列表）
        /// </summary>
        public const string LIST_SCREENSHOT = "List Screenshot";

        /// <summary>
        /// 添加截图
        /// </summary>
        public const string ADD_SCREENSHOT = "Add Screenshot";

        /// <summary>
        /// 编辑截图
        /// </summary>
        public const string EDIT_SCREENSHOT = "Edit Screenshot";

        /// <summary>
        /// 删除截图
        /// </summary>
        public const string DELETE_SCREENSHOT = "Delete Screenshot";

        /// <summary>
        /// 获取视频列表（不具有该权限无法查看视频列表）
        /// </summary>
        public const string LIST_VIDEO = "List Video";

        /// <summary>
        /// 添加视频
        /// </summary>
        public const string ADD_VIDEO = "Add Video";

        /// <summary>
        /// 编辑视频
        /// </summary>
        public const string EDIT_VIDEO = "Edit Video";

        /// <summary>
        /// 删除视频
        /// </summary>
        public const string DELETE_VIDEO = "Delete Video";
    }
}

File: ShadowEditor.Server.Go/server/operating_authority.go
package server

// GetAllOperatingAuthorities returns all operating authorities.
func GetAllOperatingAuthorities() []OperatingAuthority {
	list := []OperatingAuthority{}

	for _, val := range authorities {
		list = append(list, OperatingAuthority{
			ID:   val[0],
			Name: val[1],
		})
	}

	return list
}

var authorities = [][]string{
	{"ADMINISTRATOR", Administrator},
	{"LOGIN", Login},
	{"LIST_ANIMATION", ListAnimation},
	{"ADD_ANIMATION", AddAnimation},
	{"EDIT_ANIMATION", EditAnimation},
	{"DELETE_ANIMATION", DeleteAnimation},
	{"LIST_AUDIO", ListAudio},
	{"ADD_AUDIO", AddAudio},
	{"EDIT_AUDIO", EditAudio},
	{"DELETE_AUDIO", DeleteAudio},
	{"LIST_CATEGORY", ListCategory},
	{"SAVE_CATEGORY", SaveCategory},
	{"DELETE_CATEGORY", DeleteCategory},
	{"LIST_CHARACTER", ListCharacter},
	{"EDIT_CHARACTER", EditCharacter},
	{"SAVE_CHARACTER", SaveCharacter},
	{"DELETE_CHARACTER", DeleteCharacter},
	{"LIST_MAP", ListMap},
	{"ADD_MAP", AddMap},
	{"EDIT_MAP", EditMap},
	{"DELETE_MAP", DeleteMap},
	{"LIST_MATERIAL", ListMaterial},
	{"EDIT_MATERIAL", EditMaterial},
	{"SAVE_MATERIAL", SaveMaterial},
	{"DELETE_MATERIAL", DeleteMaterial},
	{"LIST_MESH", ListMesh},
	{"ADD_MESH", AddMesh},
	{"EDIT_MESH", EditMesh},
	{"DELETE_MESH", DeleteMesh},
	{"LIST_PARTICLE", ListParticle},
	{"EDIT_PARTICLE", EditParticle},
	{"SAVE_PARTICLE", SaveParticle},
	{"DELETE_PARTICLE", DeleteParticle},
	{"LIST_PREFAB", ListPrefab},
	{"EDIT_PREFAB", EditPrefab},
	{"SAVE_PREFAB", SavePrefab},
	{"DELETE_PREFAB", DeletePrefab},
	{"EDIT_SCENE", EditScene},
	{"SAVE_SCENE", SaveScene},
	{"PUBLISH_SCENE", PublishScene},
	{"DELETE_SCENE", DeleteScene},
	{"LIST_SCREENSHOT", ListScreenshot},
	{"ADD_SCREENSHOT", AddScreenshot},
	{"EDIT_SCREENSHOT", EditScreenshot},
	{"DELETE_SCREENSHOT", DeleteScreenshot},
	{"LIST_VIDEO", ListVideo},
	{"ADD_VIDEO", AddVideo},
	{"EDIT_VIDEO", EditVideo},
	{"DELETE_VIDEO", DeleteVideo},
}

const (
	// Administrator means the user has all the authorities.
	Administrator string = "Administrator"
	// Login means the user has logined in.
	Login string = "Login"
	// ListAnimation means the user can get the animation list.
	ListAnimation string = "List Animation"
	// AddAnimation means the user can upload animation.
	AddAnimation string = "Add Animation"
	// EditAnimation means the user can edit the animation name and thumbnail he uploaded.
	EditAnimation string = "Edit Animation"
	// DeleteAnimation means the user can delete the animation he uploaded.
	DeleteAnimation string = "Delete Animation"
	// ListAudio means the user can get the audio list.
	ListAudio string = "List Audio"
	// AddAudio means the user can upload audio.
	AddAudio string = "Add Audio"
	// EditAudio means the user can edit audio name and thumbnail he uploaded.
	EditAudio string = "Edit Audio"
	// DeleteAudio means the user can delete audio he uploaded.
	DeleteAudio string = "Delete Audio"
	// ListCategory means the user can get his category list.
	ListCategory string = "List Category"
	// SaveCategory means the user can add new category.
	SaveCategory string = "Save Category"
	// DeleteCategory means the user can delete his category.
	DeleteCategory string = "Delete Category"
	// ListCharacter means the user can get the character list.
	ListCharacter string = "List Character"
	// EditCharacter means the user can edit character name and thumbnail he uploaded.
	EditCharacter string = "Edit Character"
	// SaveCharacter means the user can save new character he created.
	SaveCharacter string = "Save Character"
	// DeleteCharacter means the user can delete character he created.
	DeleteCharacter string = "Delete Character"
	// ListMap means the user can get the texture list.
	ListMap string = "List Map"
	// AddMap means the user can add new texture.
	AddMap string = "Add Map"
	// EditMap means the user can edit texture name and thumbnail he uploaded.
	EditMap string = "Edit Map"
	// DeleteMap means the user can delete the texture he uploaded.
	DeleteMap string = "Delete Map"
	// ListMaterial means the user can get the material list.
	ListMaterial string = "List Material"
	// EditMaterial means the user can edit material name and thumbnail he saved.
	EditMaterial string = "Edit Material"
	// SaveMaterial means the user can save a new material.
	SaveMaterial string = "Save Material"
	// DeleteMaterial means the user can delete the material he saved.
	DeleteMaterial string = "Delete Material"
	// ListMesh means the user can get the mesh list.
	ListMesh string = "List Mesh"
	// AddMesh means the user can upload new mesh.
	AddMesh string = "Add Mesh"
	// EditMesh means the user can edit mesh name and thumbnail he uploaded.
	EditMesh string = "Edit Mesh"
	// DeleteMesh means the user can delete mesh he uploaded.
	DeleteMesh string = "Delete Mesh"
	// ListParticle means the user can get the particle list.
	ListParticle string = "List Particle"
	// EditParticle means the user can edit particle name and thumbnail he saved.
	EditParticle string = "Edit Particle"
	// SaveParticle means the user can save new particle.
	SaveParticle string = "Save Particle"
	// DeleteParticle means the user can delete particle he saved.
	DeleteParticle string = "Delete Particle"
	// ListPrefab means the user can get prefab list.
	ListPrefab string = "List Prefab"
	// EditPrefab means the user can edit prefab name and thumbnail.
	EditPrefab string = "Edit Prefab"
	// SavePrefab means the user can save new prefab.
	SavePrefab string = "SavePrefab"
	// DeletePrefab means the user can delete prefab.
	DeletePrefab string = "Delete Prefab"
	// EditScene means the user can edit scene name and thumbnail he saved.
	EditScene string = "Edit Scene"
	// SaveScene means the user can save new scene.
	SaveScene string = "Save Scene"
	// PublishScene means the user can publish new scene to static content.
	PublishScene string = "Publish Scene"
	// DeleteScene means the user can delete scenes he saved.
	DeleteScene string = "Delete Scene"
	// ListScreenshot means the user can get the screenshot list.
	ListScreenshot string = "List Screenshot"
	// AddScreenshot means the user can save new screenshot.
	AddScreenshot string = "Add Screenshot"
	// EditScreenshot means the user can edit screenshot name and thumbnail he saved.
	EditScreenshot string = "Edit Screenshot"
	// DeleteScreenshot means the user can delete screenshot he saved.
	DeleteScreenshot string = "Delete Screenshot"
	// ListVideo means the user can get the video list he saved.
	ListVideo string = "List Video"
	// AddVideo means the user can record new video and upload.
	AddVideo string = "Add Video"
	// EditVideo means the user can edit video name and thumbnail he saved.
	EditVideo string = "Edit Video"
	// DeleteVideo means the user can delete video he uploaded.
	DeleteVideo string = "Delete Video"
)

// OperatingAuthority is operating authority model.
type OperatingAuthority struct {
	// ID is authority ID.
	ID string
	// Name is authority name.
	Name string
}
